CountedNMap implementation

Mr-Leshiy opened the pull request:
add `CountedStorageNMap` implementation which  is similar as `CountedStorageMap` for `StorageMap` but for the `StorageNMap`
https://github.com/paritytech/substrate/issues/10602

polkadot address: 16GBceb9xiWB1R5YQGRfEYDZbWcatmLz8MPd6KZYpy4qfV7B

Mr-Leshiy committed:
add initial CountedDoubleMap implementation

Mr-Leshiy committed:
extend CountedDoubleMap functionality

Mr-Leshiy committed:
add some traits implementation for CountedStorageDoubleMap

Mr-Leshiy committed:
add basic tests for CountedStorageDoubleMap

Mr-Leshiy committed:
add mutate functions implementation

Mr-Leshiy committed:
add additional tests

Mr-Leshiy committed:
add test_option_query test

Mr-Leshiy committed:
add try_append_decode_len_works, append_decode_len_works tests

Mr-Leshiy committed:
add migrate_keys_works, translate_values tests

Mr-Leshiy committed:
add test_iter_drain_translate test

Mr-Leshiy committed:
add test_metadata test

Mr-Leshiy committed:
add remove_prefix implementation,  add test_iter_drain_prefix test

Mr-Leshiy committed:
update

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy committed:
refactor PrefixIterator usage

Mr-Leshiy committed:
Fix CI build

Mr-Leshiy committed:
fix storage_ensure_span_are_ok_wrong_gen.rs storage_ensure_span_are_ok_wrong_gen_unnamed.rs

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy commented:
@bkchr , @shawntabrizi don't want to be tactless, but can we proceed with this PR, it will be helpful to solve this issue https://github.com/paritytech/substrate/issues/9060

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

kianenigma commented:
This PR seems valuable, yet hat slipped through the crack. Partially due to lack of labeling and project assignment from our side. Hopefully will get some review soon. Thanks @Mr-Leshiy!

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

KiChjang commented:
I'm honestly interested to know why can't we just have a `CountedStorageNMap` instead? The `CountedStorageDoubleMap` can just be a type alias to a `CountedStorageNMap` that way (and the same can also be said for `CountedStorageMap`).

Mr-Leshiy commented:
@KiChjang , so sorry for the delay.
>I'm honestly interested to know why can't we just have a CountedStorageNMap instead? The CountedStorageDoubleMap can just be a type alias to a CountedStorageNMap that way (and the same can also be said for CountedStorageMap).

For my understanding it is impossible to implement in such way, because all these types of storages have different type of interfaces and also have different type of implementation. So in this case it is impossible to unify them and hide under the same implementation and interface.

KiChjang commented:
I'm sorry but this doesn't expose what exactly the impossibility is. Why is it not as simple as this:

```rust
type CountedStorageDoubleMap<P, H1, K1, H2, K2, V, Q, O, M> = CountedStorageNMap<P, (NMapKey<H1, K1>, NMapKey<H2, K2>), V, Q, O, M>;
```

What exactly is different about the interfaces?

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy commented:
@KiChjang , as a reference we can consider the following function from the original double map and from the original nmap ` contains_key`.

DoubleMap function signature [contains_key](https://github.com/Mr-Leshiy/substrate/blob/a3aa0e8f7bc8b10b981eb634f9e257f075e9aa15/frame/support/src/storage/types/double_map.rs#L178):
```
contains_key<KArg1, KArg2>(k1: KArg1, k2: KArg2) -> bool
 ```
 
 Nmap function signature [contains_key](https://github.com/Mr-Leshiy/substrate/blob/a3aa0e8f7bc8b10b981eb634f9e257f075e9aa15/frame/support/src/storage/types/nmap.rs#L125):
 ```
 contains_key<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter>(key: KArg) -> bool
 ```

KiChjang commented:
Okay, then we don't do a type alias, but call into the NMap methods underneath, i.e.

```rust
fn contains_key<KArg1, KArg2>(k1: KArg1, k2: KArg2) -> bool {
    NMap::contains_key(((k1, Hasher1::hasher()), (k2, Hasher2::hasher())))
}
```

The point here is to reduce the amount of code that we have to maintain by using NMap behind the scenes.

Mr-Leshiy commented:
@KiChjang , But how it will helps if I will replace an underneath implementation from the native DoubleMap storage to the NMap ? And if so I am coming with the general question why we have to different implementations of `DoubleMap` and `NMap` ?

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy commented:
@KiChjang I gotcha, sorry that was so annoying with questions and doubts, but I was not totally understood what do you mean and why need to change it 😊
So, what do you think if I will change this implementation just for the `NMap` ? And later will figure out how to extend it to be compatible with the DoubleMap interface ?

Mr-Leshiy committed:
add counted_nmap implementation

KiChjang commented:
That sounds like a workable plan to me. It may also be easier if you attempt to convert the existing `StorageDoubleMap` to use `StorageNMap` underneath first, but having a `CountedNMap` implementation right now is also valuable as well.

Mr-Leshiy commented:
In progress...

Mr-Leshiy committed:
add tests, fixes

Mr-Leshiy committed:
remove counted double map impl

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1753993

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1753994

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1753995

Mr-Leshiy committed:
fix metadata checks

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1755190

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1755320

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1755319

KiChjang commented:
Hmm... since we still have `StorageDoubleMap` nowadays, I think we do have to add a `CountedStorageDoubleMap` after all, since that is what people would expect, but we can then have it actually use a `CountedStorageNMap` underneath.

Mr-Leshiy commented:
@KiChjang , totally agree with you. But I would prefer to make `StorageDoubleMap` in the separate PR. But if you think that ti should be done here, I can do it, but will need some time for it :)

KiChjang reviewed (approved):
is gud.

KiChjang commented:
> @KiChjang , totally agree with you. But I would prefer to make `StorageDoubleMap` in the separate PR. But if you think that ti should be done here, I can do it, but will need some time for it :)

Totally fine as well. That also reminds me -- `CountedStorageMap` itself can use `StorageNMap` underneath too, no reason to have separate implementations of it.

shawntabrizi commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1427 @@
+		key: KArg,
+		query: QueryKind::Query,
+	) {
+		<Self as MapWrapper>::Map::set(key, query)
```
maybe it is handled inside this set function, but was expecting to see the counter increment here

KiChjang replied:
Counter should not be incremented if `key` already exists; but if it doesn't, then an increment should happen.

shawntabrizi commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1427 @@
+	where
+		Key: HasKeyPrefix<KP>,
+	{
+		let mut on_remove = 0;
+		Self::iter_prefix_values(partial_key.clone()).for_each(|_| {
+			on_remove += 1;
+		});
+		CounterFor::<Prefix>::mutate(|value| {
+			*value = value.saturating_sub(on_remove);
+		});
```
how is `limit` being respected here?

KiChjang replied:
What really should happen is that the `MultiRemovalResults` from `clear_prefix` should be looked into to check and see what the actual number of deleted items are; we then update the counter accordingly.

Mr-Leshiy replied:
updated

shawntabrizi commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1427 @@
+	/// operating on the same map should always pass `Some`, and this should be equal to the
+	/// previous call result's `maybe_cursor` field.
+	pub fn clear(limit: u32, maybe_cursor: Option<&[u8]>) -> sp_io::MultiRemovalResults {
+		CounterFor::<Prefix>::set(0u32);
+		<Self as MapWrapper>::Map::clear(limit, maybe_cursor)
```
how is limit respected here?

Mr-Leshiy replied:
updated

shawntabrizi commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1427 @@
+	/// operating on the same map should always pass `Some`, and this should be equal to the
+	/// previous call result's `maybe_cursor` field.
+	pub fn clear(limit: u32, maybe_cursor: Option<&[u8]>) -> sp_io::MultiRemovalResults {
+		CounterFor::<Prefix>::set(0u32);
```
rather than set 0, we should always remove the storage?

KiChjang replied:
Wait ok, it's not correct to always set this to 0. What should happen instead is that we get the current counter value, do the deletion, and figure out just how many items were actually deleted, then set the counter value to `current - num_deleted`.

Mr-Leshiy replied:
updated

KiChjang reviewed (changes requested):
Apparently does require changes

Mr-Leshiy committed:
update clear func

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy committed:
fix clear, clear with prefix

Mr-Leshiy committed:
fix set function

KiChjang commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1443 @@
+		KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter + Clone,
+		VArg: EncodeLike<Value>,
+	{
+		if !<Self as MapWrapper>::Map::contains_key(key.clone()) {
```
We should not clone here, instead we can just use `codec::Ref::from(&key)`. We should also eliminate the `Clone` bound on `KArg`.

KiChjang commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1443 @@
+
+	/// Remove the value under the given keys.
+	pub fn remove<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter + Clone>(key: KArg) {
+		if <Self as MapWrapper>::Map::contains_key(key.clone()) {
```
Same here, remove the `Clone` trait bound.

KiChjang commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1443 @@
+		EncodeLikeItem: EncodeLike<Item>,
+		Value: StorageAppend<Item>,
+	{
+		if !<Self as MapWrapper>::Map::contains_key(key.clone()) {
```
Same here, remove the `Clone` trait bound.

KiChjang commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1443 @@
+	pub fn clear(limit: u32, maybe_cursor: Option<&[u8]>) -> sp_io::MultiRemovalResults {
+		let result = <Self as MapWrapper>::Map::clear(limit, maybe_cursor);
+		match result.maybe_cursor {
+			None => CounterFor::<Prefix>::kill(),
```
Are you certain that `result.maybe_cursor` being `None` means there aren't any items in the map? IIUC `None` can also mean that the underlying map implementation doesn't support cursors.

KiChjang replied:
Okay nevermind, `None` does always indicate that everything was removed.

KiChjang commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1443 @@
+		key: KArg,
+	) -> QueryKind::Query {
+		let removed_value =
+			<Self as MapWrapper>::Map::mutate_exists(key, |value| core::mem::replace(value, None));
```
`mem::replace` here doesn't really remove the value from storage, does it? It just puts a `None` in its place without cleaning the entry up.

Mr-Leshiy replied:
As i understood correctly `mutate_exists` after we replaced value with `None` will clean it from the storage

Mr-Leshiy committed:
update

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
The job name - cargo-check-benches
The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1980352

KiChjang commented on frame/support/src/storage/types/key.rs:
```diff
@@ -196,6 +196,9 @@ impl_encode_like_tuples!(A, B, C, D, E, F, G, H, I, J, K, L, M, O, P);
 impl_encode_like_tuples!(A, B, C, D, E, F, G, H, I, J, K, L, M, O, P, Q);
 impl_encode_like_tuples!(A, B, C, D, E, F, G, H, I, J, K, L, M, O, P, Q, R);
 
+impl<'a, T: EncodeLike<U>, U: Encode> EncodeLikeTuple<U> for codec::Ref<'a, T, U> {}
```
The name of this trait should already give you a hint that you're only supposed to implement it on tuples. Why is this being implemented on `U`?

Mr-Leshiy replied:
I am trying to implement your suggestion https://github.com/paritytech/substrate/pull/10621#discussion_r985419556, for it is needed implement this trait also for codec::Ref

Mr-Leshiy replied:
Because `contains_key` method which I am trying to invoke here have the following trait bounds
`pub fn contains_key<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter>(key: KArg) -> bool`
https://github.com/Mr-Leshiy/substrate/blob/09583a983add3fc443a408ca0670e6112d570905/frame/support/src/storage/types/nmap.rs#L125

Mr-Leshiy replied:
The same for other methods where it is relevant to replace it with `Ref::from(&key)`

KiChjang replied:
This is really not the correct way to do so. See the comment on `EncodeLikeTuple`:

```rust
/// Marker trait to indicate that each element in the tuple encodes like the corresponding element
/// in another tuple.
///
/// This trait is sealed.
pub trait EncodeLikeTuple<T>: crate::storage::private::Sealed {}
```

There's no guarantee that T or U in your implementation here are tuples. If you want to implement `EncodeLikeTuple` for `codec::Ref`, then you can refer to the `impl_encode_like_tuples` macro, or even add additional implementations inside of it.

Mr-Leshiy committed:
final fix

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2060150

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2060151

Mr-Leshiy commented:
Still in progress ...

melekes commented on frame/support/src/storage/types/counted_map.rs:
```diff
@@ -133,8 +133,12 @@ where
 	}
 
 	/// Store or remove the value to be associated with `key` so that `get` returns the `query`.
```
should we update the comment as fn no longer "*just* sets the key"?

Mr-Leshiy replied:
Sorry, haven't got what do you mean, can you elaborate ?
As I an see in the original implementation of the `StorageMap` we have the same description https://github.com/Mr-Leshiy/substrate/blob/2a5c7781c16b845e59c4d269b51f3f99228676e6/frame/support/src/storage/mod.rs#L174

melekes replied:
I meant that we should probably say "It decrements the counter when the value is removed"

melekes commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1445 @@
+use sp_std::prelude::*;
+
+/// A wrapper around a `StorageNMap` and a `StorageValue<Value=u32>` to keep track of how many items
+/// are in a map, without needing to iterate all the values.
```
```suggestion
/// are in a map, without needing to iterate over all of the values.
```

melekes commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1445 @@
+/// A wrapper around a `StorageNMap` and a `StorageValue<Value=u32>` to keep track of how many items
+/// are in a map, without needing to iterate all the values.
+///
+/// This storage item has additional storage read and write overhead when manipulating values
```
```suggestion
/// This storage item has some additional storage read and write overhead when manipulating values
```

melekes commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1445 @@
+		KArg: EncodeLikeTuple<Key::KArg> + EncodeLike<Key::KArg> + TupleToEncodedIter,
+		VArg: EncodeLike<Value>,
+	{
+		if !<Self as MapWrapper>::Map::contains_key(Ref::from(&key)) {
```
doesn't insert return any kind of flag which indicates if value was present before?

Mr-Leshiy replied:
I have followed an original `NMap` interface, and we don't have any return value there. I don't know actually which approach should be better.  If you are insist that we should return a flag, don't see any problem to do it, I will update it. So what do think, should I fix it ?

https://github.com/Mr-Leshiy/substrate/blob/2a5c7781c16b845e59c4d269b51f3f99228676e6/frame/support/src/storage/types/nmap.rs#L171

melekes replied:
That's unfortunate. If `NMap::insert` had a flag indicating if key was present before (by analogy with https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert), then we wouldn't need to call `contains_key`.

I think we should consider changing the API, but this can be a separate issue/PR.

Mr-Leshiy replied:
I agree with you, it would be better to update an interface. But I would rather describe it in the issue first and update it in the separate PR, because it will needed to modify and revise other parts of the Storages implementation. Would you create it, or I should do it ?

melekes replied:
I'll open an issue.

melekes commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1445 @@
+	pub fn remove<KArg: EncodeLikeTuple<Key::KArg> + EncodeLike<Key::KArg> + TupleToEncodedIter>(
+		key: KArg,
+	) {
+		if <Self as MapWrapper>::Map::contains_key(Ref::from(&key)) {
```
same question about remove

melekes commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1445 @@
+	///
+	/// By returning `None` from `f` for an element, you'll remove it from the map.
+	///
+	/// NOTE: If a value fail to decode because storage is corrupted then it is skipped.
```
```suggestion
	/// NOTE: If a value can't be decoded because the storage is corrupted, then it is skipped.
```

melekes reviewed (commented):
![](https://media.giphy.com/media/NEvPzZ8bd1V4Y/giphy-downsized.gif)

Mr-Leshiy committed:
Update frame/support/src/storage/types/counted_nmap.rs

Co-authored-by: Anton <anton.kalyaev@gmail.com>

Mr-Leshiy committed:
Update frame/support/src/storage/types/counted_nmap.rs

Co-authored-by: Anton <anton.kalyaev@gmail.com>

Mr-Leshiy committed:
Update frame/support/src/storage/types/counted_nmap.rs

Co-authored-by: Anton <anton.kalyaev@gmail.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2206752

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2206825

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2206826

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy committed:
fix comments

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

KiChjang reviewed (changes requested):
We need to ensure that `EncodeLikeTuples` aren't given a blanket implementation on types that are not tuples.

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy committed:
fix suggestion

Mr-Leshiy committed:
cargo update

Mr-Leshiy commented:
@KiChjang I have updated and seems fixed with your suggestion. I have tried your solution but unfortunately it does not work.
I have implemented it as follows.
```
impl<'a, T: EncodeLike<U> + EncodeLikeTuple<U>, U: Encode> EncodeLikeTuple<U>
	for codec::Ref<'a, T, U>
{
}
impl<'a, T: EncodeLike<U> + EncodeLikeTuple<U>, U: Encode> crate::storage::private::Sealed
	for codec::Ref<'a, T, U>
{
}
```
I have tried different approaches but seems it would be the best one. What do you think ?

KiChjang committed:
Relocate impl of Sealed for Ref to module root

KiChjang commented:
Made a small relocation change, but otherwise everything LGTM.

ruseinov commented:
Before this can be merged we'll likely also need to add `storage_alias` macro support for this map along with some tests, similar to how it's done here: https://github.com/paritytech/substrate/pull/13366
Either that or a new issue, though it might be a nice precedent for posterity to have it all in one PR.

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

thiolliere commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1446 @@
+
+impl<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
+	crate::storage::generator::StorageNMap<Key, Value>
+	for CountedStorageNMap<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
```
I think this is not correct: see this example fails:
```diff
diff --git a/frame/support/src/storage/types/counted_nmap.rs b/frame/support/src/storage/types/counted_nmap.rs
index 90edc7334c4..e8e92131aed 100644
--- a/frame/support/src/storage/types/counted_nmap.rs
+++ b/frame/support/src/storage/types/counted_nmap.rs
@@ -693,6 +693,36 @@ mod test {
                }
        }
 
+       #[test]
+       fn gui_test_1() {
+               type A = CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, u32, OptionQuery>;
+               type AValueQueryWithAnOnEmpty =
+                       CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, u32, ValueQuery, ADefault>;
+               type B = CountedStorageNMap<Prefix, NMapKey<Blake2_256, u16>, u32, ValueQuery>;
+               type C = CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, u8, ValueQuery>;
+               type WithLen = CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, Vec<u32>>;
+
+               TestExternalities::default().execute_with(|| {
+                       let mut k: Vec<u8> = vec![];
+                       k.extend(&twox_128(b"test"));
+                       k.extend(&twox_128(b"Foo"));
+                       k.extend(&3u16.blake2_128_concat());
+                       assert_eq!(A::hashed_key_for((&3,)).to_vec(), k);
+
+                       assert_eq!(A::contains_key((3,)), false);
+                       assert_eq!(A::get((3,)), None);
+                       assert_eq!(AValueQueryWithAnOnEmpty::get((3,)), 98);
+                       assert_eq!(A::count(), 0);
+
+                       <A as crate::storage::StorageNMap<_, _>>::insert((3,), 10);
+                       assert_eq!(A::contains_key((3,)), true);
+                       assert_eq!(A::get((3,)), Some(10));
+                       assert_eq!(AValueQueryWithAnOnEmpty::get((3,)), 10);
+                       assert_eq!(A::count(), 1);
+               })
+       }
+
+
        #[test]
        fn test_1_key() {
                type A = CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, u32, OptionQuery>;
```

When you implement the generator it means that the trait `frame_support::storage::StorageNMap` is implemented using your implementation of the generator.  The implementation of `frame_support::storage::StorageNMap` is in `frame/support/src/storage/generator/nmap.rs:177`. This implementation makes no regards to the counter.

thiolliere replied:
fixed thanks

thiolliere commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1446 @@
+}
+
+impl<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues> crate::storage::StoragePrefixedMap<Value>
+	for CountedStorageNMap<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
```
This is not correct, StoragePrefixedMap is a trait which consider that all values of the map are after one prefix. but also doesn't account for anything like the counter.

Maybe we should improve documentation of StoragePrefixedMap with something like this:
```diff
diff --git a/frame/support/src/storage/mod.rs b/frame/support/src/storage/mod.rs
index 6a88eb676ea..c0e44043bae 100644
--- a/frame/support/src/storage/mod.rs
+++ b/frame/support/src/storage/mod.rs
@@ -1178,6 +1178,13 @@ impl<T> Iterator for ChildTriePrefixIterator<T> {
 /// ```nocompile
 /// Twox128(module_prefix) ++ Twox128(storage_prefix)
 /// ```
+///
+/// # Warning
+///
+/// The trait methods default implementations can result in inconsistent operation if some
+/// information on the keys or values are stored outside the map.
+/// For instance a counter counting for values would get inconsistent if [`Self::clear`] default
+/// implementations is called.
 pub trait StoragePrefixedMap<Value: FullCodec> {
        /// Module prefix. Used for generating final key.
        fn module_prefix() -> &'static [u8];
```

IMO just don't implement this trait. similarly to CountedStorageMap.

thiolliere replied:
fixed, thanks

Mr-Leshiy commented:
Still in progress

Mr-Leshiy commented:
Still in progress

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2886734

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2886717

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2886718

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2886732

juangirini commented:
@Mr-Leshiy is there any progress on this?

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy committed:
fix StorageEntryMetadata type

Mr-Leshiy committed:
Update frame/support/src/storage/types/nmap.rs

Co-authored-by: Guillaume Yu Thiolliere <gui.thiolliere@gmail.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3018311

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3018300

Mr-Leshiy committed:
removed StorageNMap and StoragePrefixedMap traits impl

thiolliere commented on frame/support/src/storage/types/counted_nmap.rs:
```diff
@@ -0,0 +1,1392 @@
+	type CounterPrefix: StorageInstance;
+}
+
+// Private helper trait to access map from counted storage double map
```
```suggestion
// Private helper trait to access map from counted storage n-map
```

thiolliere commented:
for convenience and boilerplate when declaring a storage in a pallet we usually declare like this:
```rust
	#[pallet::storage]
	pub(super) type Dummy2<T: Config> = CountedStorageNMap<_, T::Balance>;
```
And the macro will create the type to replace the first type parameter `_`.

So this need to be handle in the pallet macro to be usable in a pallet.

Mr-Leshiy committed:
fix tests

aaronbassett commented:
Should this also contain the `counter_storage_final_key` and `map_storage_final_prefix` methods to keep it consistent with [CountedStorageMap](https://github.com/paritytech/substrate/blob/687ddfadfe61985ab94c852ac99c0b4794ca8391/frame/support/src/storage/types/counted_map.rs#L102-L111)?

Mr-Leshiy committed:
Update frame/support/src/storage/types/counted_nmap.rs

Co-authored-by: Guillaume Yu Thiolliere <gui.thiolliere@gmail.com>

Mr-Leshiy commented:
> And the macro will create the type to replace the first type parameter _.
So this need to be handle in the pallet macro to be usable in a pallet.


@thiolliere does it requires any changes in this PR ?
If so, I am correct that it is needed to modify `#[pallet::storage]` macro to handle desired behavior ?

Mr-Leshiy commented:
@aaronbassett not sure, because I have found only one usage of these methods for the `CountedMap`
https://github.com/Mr-Leshiy/substrate/blob/687ddfadfe61985ab94c852ac99c0b4794ca8391/frame/bags-list/remote-tests/src/snapshot.rs#L25 . 

But maybe someone else can elaborate on this one

thiolliere commented:
> > And the macro will create the type to replace the first type parameter _.
> > So this need to be handle in the pallet macro to be usable in a pallet.
> 
> @thiolliere does it requires any changes in this PR ? If so, I am correct that it is needed to modify `#[pallet::storage]` macro to handle desired behavior ?

I think it is better to put in the same PR as types those are specifically designed for usage in pallets.

also macro does allow to give named type parameter, maybe only those 2 files need to be touched:
frame/support/procedural/src/pallet/parse/storage.rs
frame/support/procedural/src/pallet/expand/storage.rs

> @aaronbassett not sure, because I have found only one usage of these methods for the `CountedMap` https://github.com/Mr-Leshiy/substrate/blob/687ddfadfe61985ab94c852ac99c0b4794ca8391/frame/bags-list/remote-tests/src/snapshot.rs#L25 .
> 
> But maybe someone else can elaborate on this one

it seems those kind of method are sometime used in tests, it can be useful to ignore some storage cost in benchmarking also.

Mr-Leshiy committed:
extend pallet::storage macro with CountedStorageNMap usage

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3039116

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3039131

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3041888

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3041874

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3041873

Mr-Leshiy committed:
fix

Mr-Leshiy commented:
@thiolliere done ! 
But I don't know how it is possible to test this `#[pallet::storage]` macro :( unfortunately have not found any related examples

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3045793

thiolliere commented:
> @thiolliere done ! 
> But I don't know how it is possible to test this `#[pallet::storage]` macro :( unfortunately have not found any related examples

tests are in frame/support/test/tests/pallet.rs mostly

sam0x17 commented on frame/support/procedural/src/pallet/parse/storage.rs:
```diff
@@ -612,6 +618,16 @@ fn process_unnamed_generics(
 			)
 		},
+		StorageKind::CountedNMap => {
+			let keygen = retrieve_arg(1)?;
```
just flagging that merely having this word in a text file can cause some AV scanners to flag it so might want to rename this variable 😆

Mr-Leshiy replied:
I see your point, but not sure that renaming in one place would solve the problem, otherwise it is needed to rename in all places, not sure that it is a good idea in case of this PR

Mr-Leshiy committed:
add tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3094970

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3094971

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3094969

Mr-Leshiy committed:
fix

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3096258

Mr-Leshiy committed:
fix

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3098444

aaronbassett commented:
> @aaronbassett not sure, because I have found only one usage of these methods for the `CountedMap` https://github.com/Mr-Leshiy/substrate/blob/687ddfadfe61985ab94c852ac99c0b4794ca8391/frame/bags-list/remote-tests/src/snapshot.rs#L25 .
> 
> But maybe someone else can elaborate on this one

Only one usage of these methods _within the Substrate codebase._ It's impossible to know if developers are using them in other projects. To ensure the best possible developer experience, we should strive to provide a consistent interface with as few surprises as possible.

But, if we believe those methods on `CountedMap` are not used and provide no benefit, they should be deprecated and removed. However, if that is the case, then they should either be implemented in `CountedNMap` and then deprecated alongside the `CountedMap` methods, or this PR should not be merged until after the deprecation process has finished and the methods have been removed from `CountedMap`.

I don't know of any plans to deprecate the methods on `CountedMap`, and I think `CountedNMap` could be a really useful addition, so I would much rather it was the former.

jasl commented:
> > @aaronbassett not sure, because I have found only one usage of these methods for the `CountedMap` https://github.com/Mr-Leshiy/substrate/blob/687ddfadfe61985ab94c852ac99c0b4794ca8391/frame/bags-list/remote-tests/src/snapshot.rs#L25 .
> > But maybe someone else can elaborate on this one
> 
> @Mr-Leshiy only one usage of these methods _within the Substrate codebase._ It's impossible to know if developers are using them in other projects. To ensure the best possible developer experience, we should strive to provide a consistent interface with as few surprises as possible.
> 
> But, if we believe those methods on `CountedMap` are not used and provide no benefit, they should be deprecated and removed. However, if that is the case, then they should either be implemented in `CountedNMap` and then deprecated alongside the `CountedMap` methods, or this PR should not be merged until after the deprecation process has finished and the methods have been removed from `CountedMap`.
> 
> I don't know of any plans to deprecate the methods on `CountedMap`, and I think `CountedNMap` could be a really useful addition, so I would much rather it was the former.

This is useful. However, I don't know if using counters is an anti-pattern for blockchain.

In my use case, say I have a structured Collection and Items.
I want to store how many items are in the collection,
I need to maintain a counter, which I may use to limit the max items per collection. so please don't deprecate `CountedStorageMap`

And I need to build indexes, for example index users' items in a collection.
I will do like this
```
#[pallet::storage]
pub type Account<T: Config<I>, I: 'static = ()> = StorageNMap<
	_,
	(
		NMapKey<Blake2_128Concat, T::AccountId>, // owner
		NMapKey<Blake2_128Concat, T::CollectionId>,
		NMapKey<Blake2_128Concat, T::ItemId>,
	),
	(),
	OptionQuery,
>;
```
(This piece of code is from https://github.com/paritytech/substrate/blob/master/frame/nfts/src/lib.rs#L219-L228 )

Further, I want to limit the max items of a user in a collection, so if I have CountedNMap, that's helpful.

Manual maintenance counter fields need very careful or easy to make bugs like the counter overflow, and the code will become complex and hard to read.

I'm very +1 for the feature!

jasl commented:
> only one usage of these methods within the Substrate codebase.

<img width="736" alt="image" src="https://github.com/paritytech/substrate/assets/1024162/2ed6a0bc-55f2-41e2-93e8-f5547538052f">

And there're 10 usages of `CountedStorageMap` in Substrate codebase, not only one

aaronbassett commented:
@jasl nobody is talking about deprecating CountedStorageMap, only either adding the `counter_storage_final_key` and `map_storage_final_prefix` methods to CountedNMap or removing those methods from CountedStorageMap so that there is consistency.

jasl commented:
> @jasl nobody is talking about deprecating CountedStorageMap, only either adding the `counter_storage_final_key` and `map_storage_final_prefix` methods to CountedNMap or removing those methods from CountedStorageMap so that there is consistency.

Ok, sorry for my misunderstanding

Mr-Leshiy committed:
Add counter_storage_final_key(), map_storage_final_prefix() functions

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3151119

Mr-Leshiy committed:
update tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3152547

Mr-Leshiy committed:
fix

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3152630

Mr-Leshiy committed:
fix

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3153158

Mr-Leshiy committed:
fix

Mr-Leshiy commented:
@thiolliere done, I have added and updated tests

juangirini commented:
bot fmt

thiolliere commented on frame/support/test/tests/pallet.rs:
```diff
@@ -368,6 +368,27 @@ pub mod pallet {
 	>;
 
+	#[pallet::storage]
+	#[pallet::getter(fn counted_nmap)]
```
maybe the getter could have a test too

Mr-Leshiy replied:
done

gui1117 reviewed (approved):
Looks good to me :+1:

Mr-Leshiy committed:
update tests

Mr-Leshiy commented on frame/support/test/tests/pallet.rs:
```diff
@@ -1136,6 +1166,30 @@ fn storage_expand() {
+		k.extend(1u8.using_encoded(blake2_128_concat));
+		assert_eq!(unhashed::get::<u32>(&k), Some(3u32));
+		assert_eq!(pallet::CountedNMap::<Runtime>::count(), 1);
+
```
done

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3195167

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

jasl commented:
kindly ping

juangirini commented:
bot fmt

Mr-Leshiy committed:
Merge branch 'master' into counted-double-map

Mr-Leshiy committed:
fix fmt

Mr-Leshiy committed:
fix fmt

jasl commented:
@juangirini 
Sorry for bothering you, but is this mergeable now?

juangirini commented:
@sam0x17 @KiChjang There were some changes since your las reviews, would you give it another look before merging?

jasl commented:
Could it be merged? this is really an awesome feature

juangirini commented:
bot merge

juangirini commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into counted-double-map

juangirini commented:
bot merge

kianenigma commented:
/tip medium

kianenigma commented:
@Mr-Leshiy interested in a tip?

Mr-Leshiy commented:
@kianenigma would be great :) 
`16GBceb9xiWB1R5YQGRfEYDZbWcatmLz8MPd6KZYpy4qfV7B`

juangirini commented:
/tip medium
